wallet/update.go

```diff
@@ -36,98 +45,207 @@ type (
 	}
 )
 
-// addressPayoutEvents is a helper to add all payout transactions from an
-// apply update to a slice of transactions.
-func addressPayoutEvents(addr types.Address, cau chain.ApplyUpdate) (events []Event) {
-	index := cau.State.Index
-	state := cau.State
+// appliedEvents returns a slice of events that are relevant to the wallet
+// in the chain update.
+func appliedEvents(cau chain.ApplyUpdate, walletAddress types.Address) (events []Event) {
+	cs := cau.State
 	block := cau.Block
+	index := cs.Index
+	siacoinElements := make(map[types.SiacoinOutputID]types.SiacoinElement)
 
-	// cache the source of new immature outputs to show payout transactions
-	if state.FoundationPrimaryAddress == addr {
-		events = append(events, Event{
-			ID:    types.Hash256(index.ID.FoundationOutputID()),
-			Index: index,
-			Transaction: types.Transaction{
-				SiacoinOutputs: []types.SiacoinOutput{
-					state.FoundationSubsidy(),
-				},
-			},
-			Inflow:    state.FoundationSubsidy().Value,
-			Source:    EventSourceFoundationPayout,
+	// cache the value of spent siacoin elements to use when calculating outflow
+	cau.ForEachSiacoinElement(func(se types.SiacoinElement, spent bool) {
+		if spent {
+			siacoinElements[types.SiacoinOutputID(se.ID)] = se
+		}
+	})
+
+	addEvent := func(id types.Hash256, data any) {
+		ev := Event{
+			ID:        id,
+			Index:     index,
+			Data:      data,
 			Timestamp: block.Timestamp,
-		})
+		}
+
+		switch data := data.(type) {
+		case EventMinerPayout:
+			ev.Inflow = data.SiacoinElement.SiacoinOutput.Value
+			ev.Type = EventTypeMinerPayout
+			ev.MaturityHeight = cs.MaturityHeight()
+		case EventFoundationSubsidy:
+			ev.Inflow = data.SiacoinElement.SiacoinOutput.Value
+			ev.Type = EventTypeFoundationSubsidy
+			ev.MaturityHeight = cs.MaturityHeight()
```

Comment:
`nit` might as well make this a convenience variable same as we do with the index